### PR TITLE
DynamoDB Item

### PR DESCRIPTION
**Pull request by kddejong**
*Issue #, if available:*

*Description of changes:*
- Add a resource for creating records in a DynamoDB table

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: resources/DynamoDB_Item/README.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# AwsCommunity::DynamoDB::Item
```
Comment (ericzbeard) [inline resources/DynamoDB_Item/README.md:1]: Readme needs to be done.

File: resources/DynamoDB_Item/inputs/inputs_1_create.json [hunk 2]
```diff
@@ -0,0 +1,12 @@
+{
+    "TableName": "test",
+    "Item": {
+        "title":  {
+            "S": "one"}
```
Comment (ericzbeard) [inline resources/DynamoDB_Item/inputs/inputs_1_create.json:5]:
```suggestion
            "S": "one"
        }
```

File: resources/DynamoDB_Item/inputs/inputs_1_invalid.json [hunk 3]
```diff
@@ -0,0 +1,12 @@
+{
+    "TableName": "test",
+    "Item": {
+        "title":  {
+            "S": "one"}
```
Comment (ericzbeard) [inline resources/DynamoDB_Item/inputs/inputs_1_invalid.json:5]:
```suggestion
            "S": "one"
        }
```

File: resources/DynamoDB_Item/inputs/inputs_1_update.json [hunk 4]
```diff
@@ -0,0 +1,12 @@
+{
+    "TableName": "test",
+    "Item": {
+        "title":  {
+            "S": "one"}
```
Comment (ericzbeard) [inline resources/DynamoDB_Item/inputs/inputs_1_update.json:5]:
```suggestion
            "S": "one"
        }
```

File: resources/DynamoDB_Item/src/awscommunity_dynamodb_item/handlers.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+                client.put_item(
+                    TableName=model.TableName,
+                    Item=item,
+                    ConditionExpression=condition
+                )
+        # Setting Status to success will signal to cfn that the operation is complete
+        progress.status = OperationStatus.SUCCESS
+    except ClientError as e:
+        if e.response['Error']['Code']=='ConditionalCheckFailedException':
+            return ProgressEvent.failed(HandlerErrorCode.AlreadyExists, f"item already exists")
```
Comment (ericzbeard) [inline resources/DynamoDB_Item/src/awscommunity_dynamodb_item/handlers.py:64]: Are we expecting that these items will never be changed externally? Drift could be an issue if so.

File: resources/DynamoDB_Item/src/awscommunity_dynamodb_item/handlers.py [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+        status=OperationStatus.IN_PROGRESS,
+        resourceModel=model,
+    )
+   
+    try:
+        if model is not None:
+            item = model.Item
+            if item is None:
+               item = {}
+
```
Comment (ericzbeard) [inline resources/DynamoDB_Item/src/awscommunity_dynamodb_item/handlers.py:44]: A few comments about what we're doing here would be nice.

**Review by ericzbeard — CHANGES_REQUESTED** [review]
Need setup and integ templates for testing.

File: resources/DynamoDB_Item/src/awscommunity_dynamodb_item/handlers.py (hunk 5 above)
Comment (kddejong) [reply resources/DynamoDB_Item/src/awscommunity_dynamodb_item/handlers.py:64]: I don't expect the the PK/SK keys to change.  That would be an entirely new item.

File: resources/DynamoDB_Item/src/awscommunity_dynamodb_item/handlers.py (hunk 6 above)
Comment (kddejong) [reply resources/DynamoDB_Item/src/awscommunity_dynamodb_item/handlers.py:44]: mypy thing.  model.Item could be None which means I want to set a default value if it is.

**Comment by JobaDiniz** [comment]
This is really needed for dynamodb seed data.
Is this abandoned?

**Comment by ericzbeard** [comment]
> This is really needed for dynamodb seed data. Is this abandoned?

No it's not, we still plan to finish this up and merge it.

File: resources/DynamoDB_Item/awscommunity-dynamodb-item.json [hunk 7]
```diff
@@ -0,0 +1,117 @@
+{
+  "typeName": "AwsCommunity::DynamoDB::Item",
+  "description": "This resource will create manage the lifecycle of items in a DynamoDB tableawscommunity-dynamodb-item.json",
```
Comment (ericzbeard) [inline resources/DynamoDB_Item/awscommunity-dynamodb-item.json:3]: Create and manage?

File: resources/DynamoDB_Item/resource-role-prod.yaml [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    Properties:
+      MaxSessionDuration: 8400
+      AssumeRolePolicyDocument:
+        Version: '2012-10-17'
+        Statement:
+          - Effect: Allow
+            Principal:
+              Service: resources.cloudformation.amazonaws.com
+            Action: sts:AssumeRole
+            Condition:
```
Comment (ericzbeard) [inline resources/DynamoDB_Item/resource-role-prod.yaml:18]: We have to remove the condition in the prod role.

File: resources/DynamoDB_Item/test/integ.yml [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+          { AttributeName: "pk", KeyType: HASH },
+          { AttributeName: "sk", KeyType: HASH },
+        ]
+      AttributeDefinitions:
+        [
+          { AttributeName: "pk", AttributeType: "S" },
+          { AttributeName: "sk", AttributeType: "S" },
+        ]
+  Item:
+    Condition: IsNotProd
```
Comment (ericzbeard) [inline resources/DynamoDB_Item/test/integ.yml:34]: What's the point of the condition? Was this copied from DeleteBucketContents?

File: resources/DynamoDB_Item/awscommunity-dynamodb-item.json [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+            "$ref": "#/definitions/AttributeValue"
+          }
+        },
+        "minProperties": 1,
+        "maxProperties": 2,
+        "additionalProperties": false
+      }
+  },
+  "properties": {
+    "Item": {
```
Comment (ericzbeard) [inline resources/DynamoDB_Item/awscommunity-dynamodb-item.json:85]: Should this be a map, so multiple key-value-pairs can be added?

**Review by ericzbeard — CHANGES_REQUESTED** [review]

File: resources/DynamoDB_Item/awscommunity-dynamodb-item.json (hunk 10 above)
Comment (kddejong) [reply resources/DynamoDB_Item/awscommunity-dynamodb-item.json:117]: Item maps to Attribute which is
```
"Attribute": {
      "patternProperties": {
        "^[A-Za-z0-9_@./#&+-]{1,255}$": {
          "$ref": "#/definitions/AttributeValue"
        }
      },
      "additionalProperties": false
    },
```

This should allow for the map functionality using `patternProperties`

Comment (ericzbeard) [reply resources/DynamoDB_Item/awscommunity-dynamodb-item.json:117]: We should add multiple items in the sample/integ templates.

**Review by ericzbeard — APPROVED** [review]

**Comment by JobaDiniz** [comment]
Is this merged? How can I leverage it?

**Comment by ericzbeard** [comment]
It's merged but not published to the registry yet. If you want to use it now, you would need to clone the repo and do a private registration yourself. Actually it would be great if you could give it a try and see if it meets your needs before we publish!

**Pull request merged**